DeepSpeech.py

```diff
@@ -447,61 +447,74 @@ def train():
+                def __init__(self):
+                    progressbar.widgets.FormatLabel.__init__(self, format='Loss: %(mean_loss)f')
+
+                def __call__(self, progress, data):
+                    data['mean_loss'] = total_loss / step_count if step_count else 0.0
+                    return progressbar.widgets.FormatLabel.__call__(self, progress, data)
+
             if FLAGS.show_progressbar:
-                pbar = progressbar.ProgressBar(max_value=num_steps, redirect_stdout=True).start()
-            else:
-                pbar = lambda i: i
+                pbar = progressbar.ProgressBar(widgets=['Epoch {}'.format(epoch),
+                                                        ' | ', progressbar.widgets.Timer(),
+                                                        ' | Steps: ', progressbar.widgets.Counter(),
+                                                        ' | ', LossWidget()])
+                pbar.start()
 
             # Initialize iterator to the appropriate dataset
             session.run(init_op)
 
             # Batch loop
-            for step_index in pbar(range(num_steps)):
-                if coord.should_stop():
+            while True:
+                try:
+                    _, current_step, batch_loss, step_summary = \
+                        session.run([train_op, global_step, loss, step_summaries_op],
+                                    feed_dict=feed_dict)
+                except tf.errors.OutOfRangeError:
                     break
 
-                _, current_step, batch_loss, step_summary = \
-                    session.run([train_op, global_step, loss, step_summaries_op],
-                                feed_dict=feed_dict)
                 total_loss += batch_loss
+                step_count += 1
+
+                if FLAGS.show_progressbar:
+                    pbar.update(step_count)
+
                 step_summary_writer.add_summary(step_summary, current_step)
 
                 if is_train and FLAGS.checkpoint_secs > 0 and time.time() - checkpoint_time > FLAGS.checkpoint_secs:
                     checkpoint_saver.save(session, checkpoint_path, global_step=current_step)
                     checkpoint_time = time.time()
 
-            return total_loss / num_steps
+            if FLAGS.show_progressbar:
+                pbar.finish()
+
+            return total_loss / step_count
 
         log_info('STARTING Optimization')
         best_dev_loss = float('inf')
         dev_losses = []
-        coord = tf.train.Coordinator()
-        with coord.stop_on_exception():
+        try:
             for epoch in range(FLAGS.epochs):
-                if coord.should_stop():
-                    break
-
                 # Training
-                log_info('Training epoch %d...' % epoch)
-                train_loss = run_set('train', train_init_op, train_batches)
-                log_info('Finished training epoch %d - loss: %f' % (epoch, train_loss))
```

Comment:
There is one little problem in relying on the progress bar (and I like the beauty of the concept): The huge number of updates leads to huge (slowly downloading) logs and so I was more than one time tempted to disable it.
How about either
 - (dynamically) lowering update frequency or
 - keeping fallback logging if disabled?

Reply:
I totally overlooked that there’s a flag to disable the progress bar, I’ll keep the logging if the bar is disabled.